<a href="https://colab.research.google.com/github/ElBaron-27/UACJ_Ruben/blob/main/Actividad_Calificada_07_Datos_Duplicados_271774.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Actividad calificada: Datos duplicados**
### Detección, cuantificación y tratamiento de duplicados en un dataset real

---

| | |
|---|---|
| **Materia** | Programación para Analítica Descriptiva y Predictiva |
| **Programa** | Maestría en Inteligencia Artificial y Analítica de Datos (MIAAD) — UACJ |
| **Unidad** | 02 Análisis Descriptivo de los Datos |
| **Tema** | Limpieza y transformación de datos: duplicados |
| **Modalidad** | Individual |
| **Valor** | 100 puntos |
| **Dataset** | Hotel Booking Demand (Antonio, Almeida y Nunes, 2019) |
| **Fecha de entrega** | 28.09.2026 |

---

## Datos del estudiante

| | |
|---|---|
| **Nombre completo** |Ruben Alejandro Uribe Rodriguez |
| **Matrícula** | 271774 |

## Instrucciones

1. Trabaja en Google Colab. Guarda una copia en tu Drive (**Archivo → Guardar una copia en Drive**) antes de empezar.
2. Completa tus datos en la tabla anterior.
3. Resuelve las actividades **en orden**: algunas usan variables creadas en actividades anteriores.
4. Escribe tu código en las celdas marcadas con `# Escribe tu código aquí`. Puedes agregar celdas si lo necesitas.
5. **Comenta tu código**: cada bloque debe explicar qué hace y por qué. Los comentarios forman parte de la calificación de cada actividad.
6. Responde las preguntas en las celdas **Respuesta**, con tus propias palabras y apoyándote en los resultados que obtuviste (cita cifras).
7. Usa los métodos vistos en el notebook de la sesión: `duplicated()`, `drop_duplicates()`, `value_counts()`, `sort_values()`, `groupby()`, `pd.concat()`, máscaras booleanas, etc.
8. Antes de entregar, ejecuta todo el notebook (**Entorno de ejecución → Reiniciar y ejecutar todo**) y verifica que no haya errores.
9. **Entrega en Moodle** un archivo de texto `Actividad_Duplicados_<Matricula>.txt` que contenga el **enlace** a tu notebook (Google Colab o GitHub), con las salidas visibles.
   - Colab: **Compartir → Acceso general → Cualquier persona con el enlace (Lector)**.
   - GitHub: el repositorio debe ser **público**.
   - Opcional: puedes adjuntar también el archivo `.ipynb`.

## Contexto

El dataset **Hotel Booking Demand** contiene 119,390 reservas de dos hoteles de Portugal (un hotel urbano, *City Hotel*, y un hotel vacacional, *Resort Hotel*) con llegada entre julio de 2015 y agosto de 2017. Cada fila es una reserva y hay 32 columnas. Algunas relevantes para esta actividad:

| Columna | Descripción |
|---|---|
| `hotel` | `City Hotel` o `Resort Hotel` |
| `is_canceled` | 1 si la reserva se canceló, 0 si no |
| `lead_time` | Días entre la reserva y la fecha de llegada |
| `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month` | Fecha de llegada |
| `adults`, `children`, `babies` | Número de huéspedes |
| `country` | País de origen (código ISO de 3 letras) |
| `market_segment` | Segmento de mercado (`Online TA`, `Groups`, `Direct`, …) |
| `deposit_type` | Tipo de depósito (`No Deposit`, `Non Refund`, `Refundable`) |
| `agent`, `company` | Identificador de la agencia o empresa que hizo la reserva (con faltantes) |
| `customer_type` | Tipo de cliente (`Transient`, `Group`, …) |
| `adr` | Tarifa diaria promedio (*average daily rate*) |
| `reservation_status` | Estado final: `Check-Out`, `Canceled`, `No-Show` |
| `reservation_status_date` | Fecha del último cambio de estado |

> **Importante**: el dataset **no incluye un identificador de reserva**. Tenlo presente al interpretar los duplicados.

## Rúbrica

| Actividad | Puntos | Excelente | Bueno | Insuficiente | No entregado / Incorrecto |
|---|---|---|---|---|---|
| **1. Diagnóstico inicial** | 20 | 18–20: conteos y porcentajes correctos con los tres valores de `keep`; duplicados por hotel calculados; código comentado; conclusión con cifras. | 14–17: resultados correctos con omisiones menores en comentarios o en la conclusión. | 1–13: conteos incompletos o con errores; conclusión sin sustento. | 0: no se entregó o el código no produce resultados. |
| **2. Inspección y criterio** | 25 | 23–25: grupos inspeccionados y ordenados; tamaños de grupo analizados; comparación duplicados vs. total; argumento sólido sobre si son errores, con evidencia. | 18–22: análisis correcto pero con argumento parcial o poca evidencia. | 1–17: inspección superficial; argumento sin datos. | 0: no se entregó o es incorrecto. |
| **3. Impacto en el análisis** | 25 | 23–25: métricas por hotel antes y después con `groupby()`; tabla comparativa con `pd.concat()`; interpretación del efecto en el análisis y en un modelo predictivo. | 18–22: métricas correctas; tabla o interpretación incompletas. | 1–17: métricas con errores o sin comparación. | 0: no se entregó o es incorrecto. |
| **4. Tratamiento y verificación** | 20 | 18–20: limpieza verificada; duplicados parciales detectados e interpretados; registro de decisión completo. | 14–17: limpieza correcta; análisis parcial o registro incompleto. | 1–13: limpieza sin verificación o análisis parcial erróneo. | 0: no se entregó o es incorrecto. |
| **Reflexión final** | 10 | 9–10: respuestas fundamentadas en los resultados propios. | 7–8: respuestas correctas pero generales. | 1–6: respuestas vagas o sin relación con los resultados. | 0: no se entregó. |
| **Total** | **100** | | | | |

**Notas**:

- Las respuestas escritas de cada actividad (1.4, 2.4, 3.4, 4.2 y 4.3) y los comentarios del código cuentan para la calificación **de esa actividad**, no solo para la reflexión final.
- Entrega tardía: se descuenta **5 % por día natural** de retraso, hasta 7 días; después de ese plazo la calificación es 0. El descuento se aplica sobre la calificación obtenida.

## Carga de datos

Ejecuta la siguiente celda. No requiere modificaciones.

In [1]:
import pandas as pd

# Carga del dataset Hotel Booking Demand desde GitHub (TidyTuesday, 2020-02-11)
url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-02-11/hotels.csv'
hoteles = pd.read_csv(url)

# Dimensiones y primeras filas
print('Dimensiones:', hoteles.shape)
hoteles.head()

Dimensiones: (119390, 32)


,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,...,No Deposit,NaN,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,...,No Deposit,304.0,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,...,No Deposit,240.0,NaN,0,Transient,98.0,0,1,Check-Out,2015-07-03


---
## Actividad 1. Diagnóstico inicial (20 puntos)

### 1.1 Conteo de duplicados exactos (5 pts)

Calcula el número de filas marcadas como duplicadas en `hoteles` considerando **todas** las columnas, con `keep='first'`, `keep='last'` y `keep=False`. Imprime los tres resultados con un texto que explique qué significa cada uno.

In [2]:
df = pd.DataFrame(hoteles)

# Filas sobrantes (se conserva la primera aparición)
hoteles_first = df.duplicated(keep='first').sum()

# Filas sobrantes (se conserva la última aparición)
hoteles_last = df.duplicated(keep='last').sum()

# Todas las filas que pertenecen a algún grupo repetido
hoteles_todas = df.duplicated(keep=False).sum()

print(f'keep="first": {hoteles_first} filas sobrantes')
print(f'keep="last" : {hoteles_last} filas sobrantes')
print(f'keep=False  : {hoteles_todas} filas involucradas en duplicados')

keep="first": 31994 filas sobrantes
keep="last" : 31994 filas sobrantes
keep=False  : 40165 filas involucradas en duplicados


### 1.2 Magnitud del problema (5 pts)

Con los resultados de 1.1, calcula e imprime:

- El **porcentaje** de filas sobrantes respecto al total.
- El número de **grupos repetidos** (combinaciones de valores que aparecen más de una vez).

In [3]:
# Proporción de filas sobrantes respecto al total
hoteles_sobrantes = df.duplicated(keep='first').sum()
proporcion = df.duplicated(keep='first').mean()

# Número de grupos repetidos
frecuencias = df.value_counts(dropna=False)
hoteles_grupos_repetidos = (frecuencias > 1).sum()

print(f'Total de filas      : {len(df)}')
print(f'Filas sobrantes     : {hoteles_sobrantes}')
print(f'Porcentaje          : {proporcion * 100:.2f} %')
print(f'Grupos repetidos    : {hoteles_grupos_repetidos}')

Total de filas      : 119390
Filas sobrantes     : 31994
Porcentaje          : 26.80 %
Grupos repetidos    : 8171


### 1.3 Duplicados por hotel (5 pts)

Determina **cuántas** filas sobrantes (`keep='first'`) hay en cada hotel y qué **porcentaje** representan respecto al total de reservas de ese hotel.

Sugerencia: crea una copia de `hoteles`, agrega una columna booleana `es_duplicado` con el resultado de `duplicated()` y agrupa por `hotel`.

In [4]:
hoteles_copia = hoteles.copy()
hoteles_copia['es_duplicado'] = hoteles_copia.duplicated(keep='first')

resultado = hoteles_copia.groupby('hotel')['es_duplicado'].agg(
    filas_sobrantes='sum',
    total_reservas='count'
)

resultado['porcentaje'] = resultado['filas_sobrantes'] / resultado['total_reservas'] * 100
resultado

,filas_sobrantes,total_reservas,porcentaje
hotel,,,
City Hotel,25902,79330,32.650952
Resort Hotel,6092,40060,15.207189


### 1.4 Conclusión (5 pts)

¿Qué tan grave es el problema de duplicados en este dataset? ¿Afecta por igual a ambos hoteles? Responde con cifras.

**Respuesta 1.4:**
Es un problema grave debido a que, si consideramos el total del dataset, sobran 31,994 filas de 119,390 el cual representa alrededor del 26.8% de total, esto es, en 1 de cada 4 hay duplicados. Sin embargo, como se puede observar, esto afecta en mayor medida a City Hotel ya que tiene 32.7% de filas sobrantes mientras que el Resort Hotel tiene el 15.2%, además de que como se puede observar con las filas sobrantes, la mayor parte (81%) son de City Hotel.


---
## Actividad 2. Inspección y criterio de duplicado (25 puntos)

### 2.1 Ver los grupos repetidos (6 pts)

Filtra las filas involucradas en duplicados (`keep=False`), **ordénalas** con `sort_values()` para que las copias queden juntas y muestra las primeras 12 filas con estas columnas:

`hotel`, `lead_time`, `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month`, `adults`, `country`, `market_segment`, `deposit_type`, `adr`, `reservation_status`.

Elige tú las columnas de ordenamiento y explica tu elección en un comentario.

In [5]:
columnas_orden = [
    'hotel', 'arrival_date_year', 'arrival_date_month',
    'arrival_date_day_of_month', 'lead_time'
]

columnas_mostrar = [
    'hotel', 'lead_time', 'arrival_date_year', 'arrival_date_month',
    'arrival_date_day_of_month', 'adults', 'country', 'market_segment',
    'deposit_type', 'adr', 'reservation_status'
]

# Filas involucradas en duplicados (originales + copias)
mask = hoteles.duplicated(keep=False)
duplicadas = hoteles[mask]

# Se ordena por hotel, fecha de llegada y lead_time para que reservas
# similares queden juntas y sea más fácil inspeccionar los duplicados.
duplicadas = duplicadas.sort_values(by=columnas_orden)

# Mostrar las primeras 12 filas con las columnas solicitadas
duplicadas[columnas_mostrar].head(12)

,hotel,lead_time,arrival_date_year,arrival_date_month,arrival_date_day_of_month,adults,country,market_segment,deposit_type,adr,reservation_status
40574,City Hotel,2,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Canceled
40580,City Hotel,2,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Canceled
40588,City Hotel,2,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Canceled
40562,City Hotel,117,2015,August,1,1,ITA,Offline TA/TO,No Deposit,72.5,Check-Out
40563,City Hotel,117,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Canceled
40564,City Hotel,117,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Canceled
40566,City Hotel,117,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Check-Out
40567,City Hotel,117,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Check-Out
40568,City Hotel,117,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Canceled
40569,City Hotel,117,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Canceled


### 2.2 Tamaño de los grupos (6 pts)

Usa `value_counts()` sobre el `DataFrame` completo para contar cuántas veces aparece cada combinación de filas. Considera que las columnas `agent` y `company` tienen faltantes (revisa el parámetro `dropna`).

Con ese resultado:

1. Muestra **cuántos grupos** hay de cada tamaño (2 copias, 3 copias, …). Basta con los 10 tamaños más frecuentes.
2. Indica el **tamaño del grupo más grande** y muestra su combinación de valores.

In [6]:
# Contar cuántas veces aparece cada combinación completa de valores.
conteo = hoteles.value_counts(dropna=False)

# Solo se consideran grupos repetidos (2 o más apariciones)
grupos_repetidos = conteo[conteo >= 2]

# 1. Número de grupos de cada tamaño
grupos_por_tamano = grupos_repetidos.value_counts().head(10)

print("Grupos por tamaño:", grupos_por_tamano)

# 2. Tamaño del grupo más grande
tamano_max = conteo.max()

# Combinación de valores correspondiente al grupo más grande
combinacion = pd.Series(
    conteo.index[0],
    index=conteo.index.names
    )

print("\nTamaño del grupo más grande:", tamano_max)
print("\nCombinación de valores:")
print(combinacion)

Grupos por tamaño: count
2     5034
3      969
4      497
5      275
6      171
10     123
7      115
8      103
20      97
9       93
Name: count, dtype: int64

Tamaño del grupo más grande: 180

Combinación de valores:
hotel                             City Hotel
is_canceled                                1
lead_time                                277
arrival_date_year                       2016
arrival_date_month                  November
arrival_date_week_number                  46
arrival_date_day_of_month                  7
stays_in_weekend_nights                    1
stays_in_week_nights                       2
adults                                     2
children                                 0.0
babies                                     0
meal                                      BB
country                                  PRT
market_segment                        Groups
distribution_channel                   TA/TO
is_repeated_guest                          0
previous_cancel

### 2.3 ¿En qué se distinguen las filas duplicadas? (6 pts)

Compara las filas involucradas en duplicados (`keep=False`) con el dataset completo en:

- Distribución de `market_segment` (proporciones).
- Distribución de `deposit_type` (proporciones).
- Tasa de cancelación (`is_canceled`).

In [7]:
# Filas involucradas en duplicados (originales + copias)
dup = hoteles[hoteles.duplicated(keep=False)]

# Distribución de market_segment
seg = pd.DataFrame({
    'completo': hoteles['market_segment'].value_counts(normalize=True),
    'duplicadas': dup['market_segment'].value_counts(normalize=True)
}).fillna(0)
seg['diferencia'] = seg['duplicadas'] - seg['completo']

# Distribución de deposit_type
dep = pd.DataFrame({
    'completo': hoteles['deposit_type'].value_counts(normalize=True),
    'duplicadas': dup['deposit_type'].value_counts(normalize=True)
}).fillna(0)
dep['diferencia'] = dep['duplicadas'] - dep['completo']

# Tasa de cancelación
tasa = pd.DataFrame({
    'completo': [hoteles['is_canceled'].mean()],
    'duplicadas': [dup['is_canceled'].mean()]
}, index=['tasa_cancelacion'])

print((seg * 100).round(2))
print((dep * 100).round(2))
print((tasa * 100).round(2))

                completo  duplicadas  diferencia
market_segment                                  
Aviation            0.20        0.05       -0.15
Complementary       0.62        0.19       -0.44
Corporate           4.44        3.67       -0.77
Direct             10.56        3.50       -7.06
Groups             16.59       41.61       25.02
Offline TA/TO      20.29       30.40       10.11
Online TA          47.30       20.59      -26.72
Undefined           0.00        0.00       -0.00
              completo  duplicadas  diferencia
deposit_type                                  
No Deposit       87.65       64.17      -23.48
Non Refund       12.22       35.64       23.43
Refundable        0.14        0.19        0.05
                  completo  duplicadas
tasa_cancelacion     37.04       58.39


### 2.4 Criterio (7 pts)

El dataset no tiene identificador de reserva. Con base en la evidencia de 2.1, 2.2 y 2.3, responde:

a) ¿Las filas duplicadas son **errores de registro** o podrían ser **reservas distintas** que coinciden en todos sus valores? Argumenta con datos.

b) Las columnas `agent` y `company` tienen muchos faltantes. ¿Cómo trata `duplicated()` esos `NaN` y qué implicación tiene para este análisis?

**Respuesta 2.4:**

a) Con los datos disponibles no se puede afirmar que todas las filas duplicadas sean errores de registro. Los duplicados presentan patrones que podrían ser compatibles tanto con registros repetidos como con reservas distintas que coinciden en sus características. Por ejemplo, se concentran en Groups (41.6 % entre los duplicados frente a 16.6 % en el dataset completo) y en Non Refund (35.6 % frente a 12.2 %), y existen grupos grandes de filas idénticas, con un máximo de 180 apariciones. Dado que el dataset no contiene un identificador único de reserva, eliminar todas las filas repetidas implicaría el riesgo de eliminar reservas legítimas.

b) duplicated trata dos NaN en la misma columna como iguales. company es nula en el 94.3 % de las filas y agent en el 13.7 %, y si los NaN no contaran como iguales solo se detectarían 28 sobrantes en lugar de 31,994. Dado que NaN significa "sin agente o empresa" implica que esas columnas casi no ayudan a distinguir reservas y que dos reservas distintas pueden parecer idénticas si coinciden en el resto, lo que refuerza lo mencionado en el punto a) de que sería riesgoso eliminar duplicados debido a que algunos podrían ser reales.

---
## Actividad 3. Impacto en el análisis (25 puntos)

### 3.1 Versión sin duplicados (5 pts)

Crea `hoteles_limpio` eliminando los duplicados exactos, conservando la primera aparición y **renumerando el índice**. Imprime el número de filas antes y después y cuántas se eliminaron.

In [8]:
filas_antes = len(hoteles)

hoteles_limpio = hoteles.drop_duplicates(keep='first').reset_index(drop=True)

filas_despues = len(hoteles_limpio)

print(f"Filas antes:      {filas_antes:,}")
print(f"Filas después:    {filas_despues:,}")
print(f"Filas eliminadas: {filas_antes - filas_despues:,}")

Filas antes:      119,390
Filas después:    87,396
Filas eliminadas: 31,994


### 3.2 Métricas por hotel (8 pts)

Para `hoteles` y para `hoteles_limpio`, calcula por hotel con `groupby()`:

- Número de reservas (`.size()`).
- Tasa de cancelación (media de `is_canceled`).
- Tarifa diaria promedio (media de `adr`).

Nota: para calcular la media de varias columnas a la vez, selecciona una lista de columnas: `df.groupby('hotel')[['col1', 'col2']].mean()`.

Guarda las medias en dos `DataFrame`, `resumen_original` y `resumen_limpio`, con `hotel` como columna (usa `reset_index()`).

In [9]:
def resumen_por_hotel(df):
    # Tasa de cancelación y tarifa diaria promedio (medias de varias columnas a la vez)
    medias = df.groupby('hotel')[['is_canceled', 'adr']].mean()
    # Número de reservas por hotel
    medias['reservas'] = df.groupby('hotel').size()
    return medias.reset_index()[['hotel', 'reservas', 'is_canceled', 'adr']]

resumen_original = resumen_por_hotel(hoteles)
resumen_limpio = resumen_por_hotel(hoteles_limpio)

print(resumen_original)
print(resumen_limpio)

          hotel  reservas  is_canceled         adr
0    City Hotel     79330     0.417270  105.304465
1  Resort Hotel     40060     0.277634   94.952930
          hotel  reservas  is_canceled         adr
0    City Hotel     53428     0.300386  110.985944
1  Resort Hotel     33968     0.234809   99.025346


### 3.3 Tabla comparativa (6 pts)

Agrega a cada resumen una columna `version` (`'original'` y `'sin duplicados'`) y únelos con `pd.concat()` en un `DataFrame` llamado `comparacion`, con índice renumerado y ordenado por `hotel`, de modo que las dos versiones de cada hotel queden juntas.

In [10]:
resumen_original['version'] = 'original'
resumen_limpio['version'] = 'sin duplicados'

comparacion = (
    pd.concat([resumen_original, resumen_limpio], ignore_index=True)
      .sort_values('hotel', kind='stable')
      .reset_index(drop=True)
)

comparacion

,hotel,reservas,is_canceled,adr,version
0,City Hotel,79330,0.417270,105.304465,original
1,City Hotel,53428,0.300386,110.985944,sin duplicados
2,Resort Hotel,40060,0.277634,94.952930,original
3,Resort Hotel,33968,0.234809,99.025346,sin duplicados


### 3.4 Interpretación (6 pts)

a) ¿Qué métrica cambia más al eliminar duplicados? ¿En qué hotel el cambio es mayor? Explica por qué, relacionándolo con lo encontrado en la Actividad 2.

b) Si se entrenara un modelo para predecir `is_canceled` con los datos originales, ¿qué problemas causarían los duplicados en el entrenamiento y en la evaluación del modelo?

**Respuesta 3.4:**

a) Entre las métricas de tasa y promedio, la que presenta el cambio más marcado es la tasa de cancelación, especialmente en City Hotel. Ahí pasa de 41.7 % a 30.0 % mientras que en el Resort Hotel pasa de 27.8 % a 23.5 %. La tarifa diaria promedio también cambia, pero menos: sube de 105.30 a 110.99 en el City Hotel y de 94.95 a 99.03 en el Resort Hotel. Esto se relaciona con la Actividad 2 ya que el City Hotel concentra el 81 % de las filas sobrantes (25,902 de 31,994), y en 2.3 vimos que las filas duplicadas estaban sobrerrepresentadas en Groups y Non Refund y tenían una tasa de cancelación de 58.4 % frente a 37.0 % del dataset completo.

b) Si se entrenara un modelo para predecir is_canceled con los datos originales, los duplicados podrían causar un sesgo en el entrenamiento. Como el 26.8 % de las filas son sobrantes (31,994 de 119,390), es probable que copias idénticas de una misma reserva quedaran repartidas entre entrenamiento y prueba, y podría ocurrir que el modelo podría memorizarlas y su desempeño se vería inflado sin reflejar su capacidad de predecir reservas nuevas y el modelo quedaría mal calibrado.

---
## Actividad 4. Tratamiento y verificación (20 puntos)

### 4.1 Verificación (5 pts)

Verifica sobre `hoteles_limpio` que:

- No quedan duplicados exactos.
- No hay índices repetidos y el índice va de 0 a `len(hoteles_limpio) - 1`.

In [11]:
# 1. No quedan duplicados exactos
print("Duplicados restantes:", hoteles_limpio.duplicated().sum())

# 2. Índice sin repetidos y de 0 a len - 1
print("Índice único:", hoteles_limpio.index.is_unique)
print("Índice mínimo:", hoteles_limpio.index.min())
print("Índice máximo:", hoteles_limpio.index.max())
print("len - 1:      ", len(hoteles_limpio) - 1)

Duplicados restantes: 0
Índice único: True
Índice mínimo: 0
Índice máximo: 87395
len - 1:       87395


### 4.2 Duplicados parciales (10 pts)

La columna `reservation_status_date` registra la fecha del último cambio de estado. Detecta las filas de `hoteles_limpio` que **coinciden en todas las columnas excepto `reservation_status_date`**.

1. Cuenta cuántas filas sobrantes hay con ese criterio.
2. Muestra algunas de esas filas ordenadas, con las columnas `hotel`, `lead_time`, `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month`, `adr`, `reservation_status` y `reservation_status_date`.
3. Muestra la distribución de `reservation_status` en esas filas.

Después responde: ¿qué podrían representar estas filas? ¿Las eliminarías? Justifica.

In [12]:
# Todas las columnas excepto reservation_status_date
subset = [c for c in hoteles_limpio.columns if c != 'reservation_status_date']

# 1. Filas sobrantes (se conserva la primera aparición)
sobrantes = hoteles_limpio.duplicated(subset=subset, keep='first')
print("Filas sobrantes:", sobrantes.sum())

# 2. Filas involucradas (originales + copias), ordenadas para ver las copias juntas
involucradas = hoteles_limpio[hoteles_limpio.duplicated(subset=subset, keep=False)]

columnas = ['hotel', 'lead_time', 'arrival_date_year', 'arrival_date_month',
            'arrival_date_day_of_month', 'adr', 'reservation_status',
            'reservation_status_date']

involucradas.sort_values(by=subset)[columnas].head(12)

# 3. Distribución de reservation_status en esas filas
involucradas['reservation_status'].value_counts()

display(
    involucradas.sort_values(by=subset)[columnas].head(12)
)

display(
    involucradas['reservation_status'].value_counts()
)

Filas sobrantes: 245


,hotel,lead_time,arrival_date_year,arrival_date_month,arrival_date_day_of_month,adr,reservation_status,reservation_status_date
35572,City Hotel,9,2015,September,10,95.00,Canceled,2015-09-03
35574,City Hotel,9,2015,September,10,95.00,Canceled,2015-09-01
51267,City Hotel,12,2017,July,15,139.00,Canceled,2017-07-10
51268,City Hotel,12,2017,July,15,139.00,Canceled,2017-07-07
54170,City Hotel,14,2015,October,15,100.00,Canceled,2015-10-07
54474,City Hotel,14,2015,October,15,100.00,Canceled,2015-10-13
39425,City Hotel,14,2016,April,16,127.67,Canceled,2016-04-12
39432,City Hotel,14,2016,April,16,127.67,Canceled,2016-04-04
40080,City Hotel,14,2016,May,8,159.00,Canceled,2016-04-24
40083,City Hotel,14,2016,May,8,159.00,Canceled,2016-04-26


,count
reservation_status,
Canceled,471


**Respuesta 4.2:**

Estas filas podrían representar una misma reserva cancelada que quedó registrada más de una vez con fechas de actualización distintas, por ejemplo por reenvíos o actualizaciones del sistema. También podrían ser reservas distintas que coinciden en todos sus datos y se cancelaron en días diferentes, como habitaciones de un mismo bloque, pero que solo haya cancelaciones y que la gran mayoría de los grupos sean pares apunta más a lo primero. Por tanto yo sí las eliminaría, conservando un solo registro por grupo con base en la fecha de actualización de estado, que no describe la reserva ni la estancia, y el resto de sus 31 columnas es idéntico.

### 4.3 Registro de decisión (5 pts)

Redacta, en una celda de texto, el registro de la limpieza realizada, como lo documentarías en un proyecto real. Incluye:

- **Criterio** de duplicado utilizado (columnas comparadas).
- **Regla de conservación** (`keep`) y por qué es o no relevante en este caso.
- **Resultado** (filas antes, después y eliminadas).
- **Supuestos y limitaciones**.

**Respuesta 4.3:**

1. **Criterio de duplicado.** Se consideró duplicada una fila idéntica a otra en las 32 columnas del dataset (duplicated() sin subset). Los NaN cuentan como iguales entre sí, lo cual es determinante porque company es nula en el 94.3 % de las filas y agent en el 13.7 %. Sin esa regla solo se habrían detectado 28 sobrantes en lugar de 31,994.

2. **Regla de conservación.** Se usó keep='first', que conserva la primera aparición de cada grupo y elimina las copias posteriores. En este caso la elección es irrelevante para el contenido: al ser filas idénticas en todas las columnas, conservar la primera o la última deja exactamente los mismos valores (en 1.1 ambas dan 31,994 sobrantes), y el índice se renumera con reset_index(drop=True).

3. **Resultado.** Filas antes: 119,390. Filas después: 87,396. Filas eliminadas: 31,994 (26.8 %). Por hotel, el City Hotel pasó de 79,330 a 53,428 reservas y el Resort Hotel de 40,060 a 33,968. La tasa de cancelación bajó de 41.7 % a 30.0 % en el City Hotel y de 27.8 % a 23.5 % en el Resort Hotel. La verificación en el ejercicio 4.1 confirmó 0 duplicados exactos restantes y un índice único de 0 a 87,395.

4. **Supuestos y limitaciones.** Se asumió que dos filas idénticas en las 32 columnas corresponden a la misma reserva registrada más de una vez, pero como el dataset no tiene un identificador de reserva, este supuesto no se puede confirmar. Los duplicados se concentran en Groups y en grupos de 10 o más copias, el mayor de 180, lo que sugiere que parte de ellos podrían ser reservas legítimas registradas en bloque, así que la eliminación pudo quitar reservas reales y sesgar el análisis. Además, duplicated() trata los NaN como iguales, y como company es nula en el 94.3 % de las filas y agent en el 13.7 %, esas columnas casi no ayudan a distinguir reservas, por lo que dos reservas distintas sin agente ni empresa pueden parecer idénticas si coinciden en el resto. Por todo lo anterior, yo recomendaría reportar los resultados con y sin duplicados.

---
## Reflexión final (10 puntos)

Responde brevemente, con base en tus resultados:

1. Describe una situación en la que **no** convendría eliminar los duplicados exactos de este dataset.
2. ¿Qué información adicional (columnas) le pedirías al proveedor de los datos para tomar una decisión más segura? ¿Por qué?
3. En tus propias palabras, ¿por qué `duplicated().sum() == 0` no garantiza que un dataset esté libre de duplicados?

**Respuesta final:**

1. No convendría eliminar los duplicados exactos si el objetivo fuera medir demanda u ocupación, por ejemplo, estimar cuántas habitaciones se reservaron por fecha o los ingresos por segmento. Vimos que los duplicados se concentran en Groups y en grupos de 10 o más copias, con picos en tamaños redondos como 10 y 20. Esto sugiere bloques de habitaciones reservadas de una vez, donde cada fila idéntica sería una habitación real. Eliminarlas cambiaría la perspectiva de la demanda de grupos y también sus cancelaciones por lo que al quitarlas el análisis podría subestimar el riesgo de cancelación de las reservas grupales y no reembolsables.
2. Pediría, en primer lugar, un identificador de reserva, porque permitiría distinguir de forma directa una reserva registrada dos veces de dos reservas distintas que coinciden en todos sus valores. También pediría la fecha y hora exacta de registro de cada reserva, para ver si las copias se cargaron al mismo tiempo (lo que apuntaría a error) o en momentos distintos, esto ayudaría a separar los bloques legítimos de Groups de los errores. Por último, un identificador del lote o del sistema de origen de la carga permitiría detectar cargas masivas repetidas. Con estas columnas la decisión de eliminar dejaría de basarse en suposiciones y con mayor certeza podriamos eliminar duplicados
3. duplicated().sum() == 0 solo garantiza que no hay filas idénticas en todas las columnas, pero no que no existan duplicados tal cual En este dataset lo vimos ya que después de eliminar los duplicados exactos, la verificación dio 0 duplicados restantes, y aun así al ignorar reservation_status_date aparecieron 245 duplicados parciales que solo difieren en esa fecha. Una sola diferencia en una columna basta para que duplicated() considere distintas dos filas que representan la misma reserva. Por eso un resultado de cero solo confirma que no hay copias exactas, y hay que revisar también duplicados parciales o aproximados

---
**Fin de la actividad.** Antes de entregar: **Reiniciar y ejecutar todo**, verificar que no haya errores, guardar con las salidas visibles y comprobar que el enlace abre sin iniciar sesión.